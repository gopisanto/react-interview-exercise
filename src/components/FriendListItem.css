:local(.friendListItem) {
  list-style: none;
  padding: 20px 10px 20px 10px;
  background-color: white;
  border-bottom: 1px solid #E3E3E3;
  display: flex;
}

:local(.friendInfos) {
  flex: 1 0 auto;
  display: flex;
  align-items: center;
}

:local(.friendInfos span) {
  font-weight: bold;
}

:local(.friendActions) {
  flex: 0 0 90px;
}

:local(.btnAction), :local(.btnAction):active, :local(.btnAction):focus, :local(.btnAction):hover {
  margin-right: 5px;
  color: #5C75B0;
}

:local(.details) {
  display: inline-block;
  margin-left: 5px;
}

button:focus {
  outline: 0 !important;
}