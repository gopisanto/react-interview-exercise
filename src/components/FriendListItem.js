import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import styles from './FriendListItem.css';

const FriendListItem = ({ gender, name, id, starred, starFriend, deleteFriend }) => (
  <li className={styles.friendListItem}>
    <div className={styles.friendInfos}>
      <i className={classnames('fa fa-2x', {'fa-male': gender === 'male', 'fa-female': gender === 'female'})} aria-hidden="true" />
      <div className={styles.details}>
        <div><span>{name}</span></div>
        <div>
          <small>xx friends in common</small>
        </div>
      </div>
    </div>
    <div className={styles.friendActions}>
      <button className={`btn btn-default ${styles.btnAction}`}
              onClick={() => starFriend(id)}>
        <i className={classnames('fa', {
          'fa-star': starred,
          'fa-star-o': !starred
        })} />
      </button>
      <button className={`btn btn-default ${styles.btnAction}`}
              onClick={() => deleteFriend(id)}>
        <i className="fa fa-trash" />
      </button>
    </div>
  </li>
);

FriendListItem.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  starred: PropTypes.bool,
  gender: PropTypes.string.isRequired,
  starFriend: PropTypes.func.isRequired,
  deleteFriend: PropTypes.func.isRequired
};

export default FriendListItem