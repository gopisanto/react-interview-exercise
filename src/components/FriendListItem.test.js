import React from 'react';
import renderer from 'react-test-renderer';
import {configure, shallow, mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
import sinon from 'sinon';
import FriendListItem from './FriendListItem';

configure({ adapter: new Adapter() });

describe('<FriendListItem />', () => {
    let props = {
        gender: 'male',
        name: 'Santosh',
        id: 1,
        starred: true,
        starFriend: () => null,
        deleteFriend: () => null
    }
    it('should render', () => {
        const wrapper = renderer.create(<FriendListItem {...props} />).toJSON();

        expect(wrapper).toMatchSnapshot();
    });
    it('should call starFriend action when clicked on star of a particular friend', () => {
        const spy = sinon.spy();
        props.starFriend = spy;
        const wrapper = shallow(<FriendListItem {...props} />);
        wrapper.find('i').findWhere(node => node.hasClass('fa fa-star')).simulate('click');
        expect(spy.withArgs(props.id).calledOnce);
    });
    it('should call deleteFriend action when clicked on delete of a particular friend', () => {
        const spy = sinon.spy();
        props.deleteFriend = spy;
        const wrapper = shallow(<FriendListItem {...props} />);
        wrapper.find('i').findWhere(node => node.hasClass('fa fa-trash')).simulate('click');
        expect(spy.withArgs(props.id).calledOnce);
    });
});