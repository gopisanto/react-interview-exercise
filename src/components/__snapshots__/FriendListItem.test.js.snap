exports[`<FriendListItem /> should render 1`] = `
<li
  className={undefined}>
  <div
    className={undefined}>
    <i
      aria-hidden="true"
      className="fa fa-2x fa-male" />
    <div
      className={undefined}>
      <div>
        <span>
          Santosh
        </span>
      </div>
      <div>
        <small>
          xx friends in common
        </small>
      </div>
    </div>
  </div>
  <div
    className={undefined}>
    <button
      className="btn btn-default undefined"
      onClick={[Function]}>
      <i
        className="fa fa-star" />
    </button>
    <button
      className="btn btn-default undefined"
      onClick={[Function]}>
      <i
        className="fa fa-trash" />
    </button>
  </div>
</li>
`;
